examples/features/nse-composition/README.md

```diff
@@ -0,0 +1,313 @@
+      nodeSelector:
+        kubernetes.io/hostname: ${NODE}
+EOF
+```
+
+Create NSE patch:
+```bash
+cat > patch-nse.yaml <<EOF
+---
+apiVersion: apps/v1
+kind: Deployment
+metadata:
+  name: nse-kernel
+spec:
+  template:
+    spec:
+      containers:
+        - name: nse
+          env:
+            - name: NSE_CIDR_PREFIX
+              value: 172.16.1.100/31
+            - name: NSE_SERVICE_NAME
+              value: "nse-composition"
+            - name: NSE_REGISTER_SERVICE
+              value: "false"
+            - name: NSE_LABELS
+              value: "app:gateway"
+        - name: nginx
+          image: networkservicemesh/nginx
+          imagePullPolicy: IfNotPresent
+      nodeSelector:
+        kubernetes.io/hostname: ${NODE}
+EOF                                                                                        
+```
+
+Create ConfigMap
+```bash
+cat > config-file.yaml <<EOF
+---
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: vppagent-firewall-config-file
+  namespace: ${NAMESPACE}
+data:
+  config.yaml: |
+    aclrules:
+      allow icmp:
+        ispermit: 1
+        proto: 1
+        srcportoricmptypelast: 65535
+        dstportoricmpcodelast: 65535
+      allow tcp8080:
+        ispermit: 1
+        proto: 6
+        srcportoricmptypelast: 65535
+        dstportoricmpcodefirst: 8080
+        dstportoricmpcodelast: 8080
+      forbid tcp80:
+        proto: 6
+        srcportoricmptypelast: 65535
+        dstportoricmpcodefirst: 80
+        dstportoricmpcodelast: 80
+EOF
+```
+
+Create nse-composition Network Serivice
+```bash
+cat > nse-composition-ns.yaml <<EOF
+---
+apiVersion: networkservicemesh.io/v1
+kind: NetworkService
+metadata:
+  name: nse-composition
+  namespace: nsm-system
+spec:
+  payload: ETHERNET
+  name: nse-composition
+  matches:${NS}
+EOF
+```
+
+Deploy NS configuration
+```bash
+kubectl create -f ./nse-composition-ns.yaml
+```
+
+Deploy NSC and NSE:
+```bash
+kubectl apply -k .
+```
+
+Wait for applications ready:
+```bash
+kubectl wait --for=condition=ready --timeout=1m pod -l app=nsc-kernel -n ${NAMESPACE}
+```
+```bash
+kubectl wait --for=condition=ready --timeout=1m pod -l app=nse-kernel -n ${NAMESPACE}
+```
+
+Find nsc and nse pods by labels:
+```bash
+NSC=$(kubectl get pods -l app=nsc-kernel -n ${NAMESPACE} --template '{{range .items}}{{.metadata.name}}{{"\n"}}{{end}}')
+```
+```bash
+NSE=$(kubectl get pods -l app=nse-kernel -n ${NAMESPACE} --template '{{range .items}}{{.metadata.name}}{{"\n"}}{{end}}')
+```
+
+Ping from NSC to NSE:
+```bash
+kubectl exec ${NSC} -n ${NAMESPACE} -- ping -c 4 172.16.1.100
+```
+
+Check TCP Port 8080 on NSE is accessible to NSC
+```bash
+kubectl exec ${NSC} -n ${NAMESPACE} -- wget -O /dev/null --timeout 5 "172.16.1.100:8080"
+```
+
+Check TCP Port 80 on NSE is inaccessible to NSC
+```bash
+kubectl exec ${NSC} -n ${NAMESPACE} -- wget -O /dev/null --timeout 5 "172.16.1.100:80" || echo "port :80 is unavailable"
```

Comment:
This will not trigger test error if the port is accessible. If wget succeeds, then this command chunk will have 0 exit code.
```suggestion
kubectl exec ${NSC} -n ${NAMESPACE} -- wget -O /dev/null --timeout 5 "172.16.1.100:80"
if [[ 0 -eq $? ]]; then
  echo "error: port :80 is available" >&2
  false
else
  echo "success: port :80 is unavailable"
fi
```

Reply:
done